lib/rucio/core/rule.py

```diff
@@ -2978,7 +2979,12 @@ def __evaluate_did_attach(eval_did, *, session: "Session", logger=logging.log):
 
 
 @transactional_session
-def __resolve_did_to_locks_and_replicas(did, nowait=False, restrict_rses=None, source_rses=None, only_stuck=False, *, session: "Session"):
+def __resolve_did_to_locks_and_replicas(did: models.DataIdentifier, nowait: bool = False, restrict_rses: Optional[Sequence[str]] = None,
+                                        source_rses: Optional[Sequence[str]] = None, only_stuck: bool = False, *,
+                                        session: "Session") -> tuple[list[dict[str, Any]],
+                                                                     dict[tuple[str, str], models.ReplicaLock],
+                                                                     dict[tuple[str, str], Any],
+                                                                     dict[tuple[str, str], str]]:
```

Comment:
The code suggests that the hint for the *third* member of the return type should be `dict[tuple[str, str], models.RSEFileAssociation]`. Did you try it and found issues? Same comment for the similar functions that follow.